asset-properties-widget/component/asset-properties-view/asset-properties-view.component.ts

```diff
@@ -30,76 +52,278 @@ export class AssetPropertiesViewComponent implements OnInit {
-            if (
-              property.isExistingProperty ||
-              ((this.customProperties.length != 0 &&
-                this.customProperties.find((prop) => prop.id === property.id)) || this.validateComplexProperty(property))
-            ) {
-              return property;
+            this.properties = this.config.properties.filter((property) => {
+              if (
+                property.isExistingProperty ||
+                ((this.customProperties.length != 0 &&
+                  this.customProperties.find((prop) => prop.id === property.id)) || this.validateComplexProperty(property))
+              ) {
+                return property;
+              }
+            });
+           this.config.properties = cloneDeep(this.properties);
+          }else{
+            this.properties = cloneDeep(this.config.properties);
+            this.customProperties = cloneDeep(deviceProperty);
+          }
+          this.constructComplexPropertyKeys();
+          this.config.properties.forEach(property =>{
+            if(property.computed && property.active){
+             switch (property.name) {
+              case 'alarmCountToday':
+                this.getAlarmCount(this.config.device, property,this.datePipe.transform(new Date(), 'yyyy-MM-dd'));
+                break;
+              case 'alarmCount3Months':
+                this.getAlarmCount(this.config.device, property, this.datePipe.transform(new Date().setMonth(new Date().getMonth() - 3), 'yyyy-MM-dd'));
+                break;
+              case 'eventCountToday':
+                this.getEventCount(this.config.device, property,this.datePipe.transform(new Date(), 'yyyy-MM-dd'));
+                break;
+              case 'eventCount3Months':
+                this.getEventCount(this.config.device, property, this.datePipe.transform(new Date().setMonth(new Date().getMonth() - 3), 'yyyy-MM-dd'));
+                break;
+              case 'lastMeasurement':
+                this.getLastMeasurement(property);
+                break;
+              case 'lastDeviceMessage':
+                if(!this.selectedAsset.c8y_Availability || !this.selectedAsset.c8y_Availability.lastMessage){
+                 this.getLastDeviceMessage(this.config.device);
+                }
+                break;
+              case 'configurationSnapshot':
+                this.getConfigurationSnapshot(this.selectedAsset);
+                break;
+                default:
+                  break;
+              }
             }
           });
-          this.config.properties = cloneDeep(this.properties);
-          this.constructComplexPropertyKeys();
           this.handleRealtime();
+          this.isLoading = false;
         }, 1000);
       } catch (error) {
         this.isEmptyWidget = true;
       }
+    } else this.isEmptyWidget = true;
+  }
+
+  async getAlarmCount(device:IManagedObject, property:IManagedObject, dateFrom:string){
```

Comment:
This component file is doing too many things. See if this can be split logically between the component and service

Reply:
I have relocated a couple of functions to the service.